src/Store.ts

```diff
@@ -0,0 +1,392 @@
+	 * used as the prototype for all objects returned from this store. One can set
+	 * this to the Model from dmodel/Model to return Model objects, or leave this
+	 * to null if you don't want any methods to decorate the returned
+	 * objects (this can improve performance by avoiding prototype setting),
+	 */
+	Model: NewableStoreModel;
+
+	/**
+	 * One can provide a parsing function that will permit the parsing of the data. By
+	 * default we assume the provide data is a simple JavaScript array that requires
+	 * no parsing (subclass stores may provide their own default parse function)
+	 */
+	parse: (...args: any[]) => any;
+
+	/**
+	 * Stores an object
+	 *
+	 * @param object The object to store.
+	 * @param directives Additional directives for storing objects.
+	 * @returns The object that was stored, with any changes that were made by
+	 * the storage system (like generated id)
+	 */
+	put: (object: any, directives?: dstore.PutDirectives) => Promise<any>;
+
+	/**
+	 * Indicates if client-side query engine filtering should (if the store property is true)
+	 * access object properties through the get() function (enabling querying by
+	 * computed properties), or if it should (by setting this to false) use direct/raw
+	 * property access (which may more closely follow database querying style).
+	 */
+	queryAccessors: boolean =  true;
+
+	/**
+	 * Deletes an object by its identity
+	 *
+	 * @param id The identity to use to delete the object
+	 */
+	remove: (id: any) => Promise<Object>;
+
+	/**
+	 * The query operations represented by this collection
+	 */
+	queryLog: dstore.QueryLogEntry<any>[] = [];
+
+	storage: Evented;
+
+	/**
+	 * For stores that serialize data (to send to a server, for example) the stringify
+	 * function can be specified to control how objects are serialized to strings
+	 */
+	stringify: () => any;
+
+	constructor(options?: StoreArgs) {
+		super();
+
+		// perform the mixin
+		options && lang.mixin(this, options);
+		// TODO - Don't do this, it's a workaround to make inheritance from this class via
+		// declare work properly.
+		for (let name in this) {
+			let value = this[name];
+			if (Object.prototype.toString.call(value) === '[object Function]') {
+				value.nom = name;
+			}
+		}
+
+		// TODO - This is designed specifically for working with declare-based inheritance and needs to be reworked.
+		if (this.Model && this.Model.createSubclass) {
+			// we need a distinct model for each store, so we can
+			// save the reference back to this store on it.
+			// we always create a new model to be safe.
+			this.Model = this.Model.createSubclass([]).extend({
+				// give a reference back to the store for saving, etc.
+				_store: this
+			});
+		}
+
+		// the object the store can use for holding any local data or events
+		this.storage = new Evented();
+		this._initialize();
+	}
+
+	protected _initialize() {
+		const store = this;
+		if (this.autoEmitEvents) {
```

Comment:
I'm not sure this is the best place to handle `autoEmitEvents` - in `constructor` *after* calling `_initialize` and mixing in options might be better.  Otherwise subclasses would have to call `_initialize` specifically *after* overriding `autoEmitEvents`.